command/agent/config.go

```diff
@@ -747,61 +793,28 @@ func normalizeBind(addr, bind string) string {
 // is resolved and returned with the port.
 //
 // Loopback is only considered a valid advertise address in dev mode.
-func normalizeAdvertise(addr string, bind string, defport int, dev bool) (string, error) {
+func normalizeAdvertise(addr string, bind string, defport int) (string, error) {
 	if addr != "" {
 		// Default to using manually configured address
-		_, _, err := net.SplitHostPort(addr)
+		host, port, err := net.SplitHostPort(addr)
 		if err != nil {
 			if !isMissingPort(err) {
 				return "", fmt.Errorf("Error parsing advertise address %q: %v", addr, err)
 			}
-
-			// missing port, append the default
-			return net.JoinHostPort(addr, strconv.Itoa(defport)), nil
+			host = addr
+			port = strconv.Itoa(defport)
 		}
-		return addr, nil
-	}
 
-	// Fallback to bind address first, and then try resolving the local hostname
-	ips, err := net.LookupIP(bind)
-	if err != nil {
-		return "", fmt.Errorf("Error resolving bind address %q: %v", bind, err)
-	}
-
-	// Return the first unicast address
-	for _, ip := range ips {
-		if ip.IsLinkLocalUnicast() || ip.IsGlobalUnicast() {
-			return net.JoinHostPort(ip.String(), strconv.Itoa(defport)), nil
-		}
-		if ip.IsLoopback() && dev {
-			// loopback is fine for dev mode
-			return net.JoinHostPort(ip.String(), strconv.Itoa(defport)), nil
+		ipStr, err := parseSingleIPTemplate(host)
+		if err != nil {
+			return "", fmt.Errorf("Error parsing advertise address template: %v", err)
 		}
-	}
-
-	// As a last resort resolve the hostname and use it if it's not
-	// localhost (as localhost is never a sensible default)
-	host, err := os.Hostname()
-	if err != nil {
-		return "", fmt.Errorf("Unable to get hostname to set advertise address: %v", err)
-	}
 
-	ips, err = net.LookupIP(host)
-	if err != nil {
-		return "", fmt.Errorf("Error resolving hostname %q for advertise address: %v", host, err)
+		return net.JoinHostPort(ipStr, port), nil
 	}
 
-	// Return the first unicast address
-	for _, ip := range ips {
-		if ip.IsLinkLocalUnicast() || ip.IsGlobalUnicast() {
-			return net.JoinHostPort(ip.String(), strconv.Itoa(defport)), nil
-		}
-		if ip.IsLoopback() && dev {
-			// loopback is fine for dev mode
-			return net.JoinHostPort(ip.String(), strconv.Itoa(defport)), nil
-		}
-	}
-	return "", fmt.Errorf("No valid advertise addresses, please set `advertise` manually")
+	// Fallback to bind address, as it has been resolved before.
+	return net.JoinHostPort(bind, strconv.Itoa(defport)), nil
```

Comment:
@schmichael OK, so I changed the code slightly and re-added bits from the previous version, except:

* I removed the ``os.Hostname()`` and related ``net.LookupIP(host)`` as per the problems you exposed in 2. and I'm looking up the private IP address using ``GetPrivateIP`` instead.
* before that, I'm trying to use the bind address. I still need to do the ``ip.IsLinkLocalUnicast() || ip.IsGlobalUnicast()`` dance since the bind address could be something like ``localhost`` which needs to be parsed in order to use ``IsLoopback()``. I thought I could remove this part and profit from what go-sockaddr/template offers for this instead, but I'm not sure how I could prevent 127.0.0.1 to be advertised without resolving the bind address value, in which case I need to call the functions above to do the right thing...